Replace environment.yml with conda-lock.yml

sverhoeven opened the pull request:
TODO
* [x] valid pyproject.toml for `pip install -e .[dev]`


Refs #441 

To check
- [x] conda-lock lock on linux
- [x] conda-lock install on linux
- [x] `pip install -e .` should only install ewatercycle, only few additional deps should be installed
- [x] `pip install -e .[dev]` should only install ewatercycle, few additional deps should be installed
- [x] installing without conda should give instructions

sverhoeven committed:
Replace environment.yml with conda-lock.yml

TODO
* [ ] valid pyproject.toml for `pip install -e .[dev]`

sverhoeven committed:
Configure conda-lock in pyproject.toml

Currently pip does not understand pyproject.toml file

sverhoeven committed:
Switched to hatch + deleted unneeded setup.cfg and setup.py

sverhoeven committed:
Make pyproject.toml valid + add sections not migrated by hatch

Skipped flake8 config as this will be replaced by ruff

sverhoeven committed:
Merge branch 'hatch' into lock

sverhoeven committed:
Remove esmvaltool as python dep

sverhoeven committed:
Remove docs and test from sdist + mypy stricter and fix

sverhoeven committed:
Merge remote-tracking branch 'origin/hatch' into lock

sverhoeven committed:
Move mypy from precommit to ci + remove pyroma as it is not useful anymore

sverhoeven committed:
Move mypy from precommit to ci + remove pyroma as it is not useful anymore

sverhoeven committed:
Use lock in ci

sverhoeven committed:
Add env name to ci

sverhoeven committed:
Use shell for mypy

sverhoeven committed:
Use shell for mypy

sverhoeven committed:
Merge branch 'hatch' into lock

sverhoeven committed:
Exclude lock from lint

sverhoeven committed:
Merge remote-tracking branch 'origin/main' into lock

sverhoeven committed:
Use lock in docs

sverhoeven committed:
Typo

BSchilperoort commented:
I followed the instructions: 
- environment downloaded successfully
- did `pip install . --no-deps` in checked out ewatercycle repo
- ran `pytest` and all tests passed :+1: 

However, I thought dev dependencies should not be added to the lock file but it seems like they did

sverhoeven committed:
No longer need MANIFEST.in

sverhoeven committed:
Exclude dev deps during install

Unable to exclude during lock generation as `--no-dev-dependencies` moves dev deps to main cat.

sverhoeven committed:
Merge remote-tracking branch 'origin/main' into lock

sverhoeven committed:
Use lock file to install dev deps

sverhoeven committed:
Remove unused dependencies

sverhoeven committed:
Add osx-64 + lock grpc4bmi deps

sverhoeven committed:
Regenerate lock file

sverhoeven committed:
Add missing fsspec dep

sverhoeven committed:
changelog entry

sverhoeven commented:
> I followed the instructions:
> 
> * environment downloaded successfully
> * did `pip install . --no-deps` in checked out ewatercycle repo
> * ran `pytest` and all tests passed 👍
> 
> However, I thought dev dependencies should not be added to the lock file but it seems like they did

Switched to `conda-lock install --no-dev` for production install

sverhoeven commented:
Hmm, `pip install` is installing black and tokenize-rt-5.2.0  . They have been removed in https://github.com/csdms/bmi-python/pull/46 , but no new version has been released yet.

sverhoeven committed:
Add cartopy as dep

Because it is imported in this package

sverhoeven committed:
Give nicer error when esmvaltool/esmvalcore package can not be found

sverhoeven committed:
Use custom error

sverhoeven committed:
Add test for ESMValToolNotFoundError

BSchilperoort commented:
> Hmm, `pip install` is installing black and tokenize-rt-5.2.0 . They have been removed in [csdms/bmi-python#46](https://github.com/csdms/bmi-python/pull/46) , but no new version has been released yet.

For me pip says it did the following:
`Successfully installed black-24.4.2 ewatercycle-2.2.0 gitdir-1.2.7`

And with `.[dev]`:
`Installing collected packages: webencodings, wcwidth, snowballstemmer, pure-eval, ptyprocess, fastjsonschema, distlib, commonmark, virtualenv, types-requests, types-pyyaml, types-python-dateutil, traitlets, tokenize-rt, tinycss2, sphinxcontrib-serializinghtml, sphinxcontrib-qthelp, sphinxcontrib-jsmath, sphinxcontrib-htmlhelp, sphinxcontrib-devhelp, sphinxcontrib-applehelp, rpds-py, pyzmq, pytest-runner, pytest, pyproject_hooks, pygments, pycodestyle, prompt-toolkit, pexpect, parso, pandocfilters, ordered-set, nodeenv, mypy, mistune, mdurl, jupyterlab-pygments, isort, imagesize, identify, executing, docutils, decorator, coverage, cfgv, bump2version, bleach, babel, asttokens, astroid, alabaster, stack-data, sphinx, referencing, pre-commit, matplotlib-inline, markdown-it-py, jupyter-core, jedi, deepdiff, build, sphinxcontrib-jquery, sphinx-copybutton, sphinx-autoapi, rich, recommonmark, pytest-cov, jupyter-client, jsonschema-specifications, ipython, sphinx-rtd-theme, jsonschema, nbformat, nbclient, nbconvert, nbsphinx, ewatercycle`

Which is normal as it's the dev+docs dependencies.

Perhaps edit the PR text:
> pip install -e .[dev] should only install ewatercycle, no additonal deps should be installed

BSchilperoort commented on CONTRIBUTING.md:
```diff
@@ -79,22 +81,23 @@ your pull request.
-7. Wait for [GitHub
+2. If dependencies have changed then create a new [conda lock](https://conda.github.io/conda-lock/) file with ``conda-lock lock -f pyproject.toml``.
+3. Determine what new version (major, minor or patch) to use. Package uses `semantic versioning <https://semver.org>`_.
+4. Run ``bump2version <major|minor|patch>`` to update version in package files.
```
This shouldn't be here anymore right (bump2version)?

BSchilperoort replied:
Oh wait it's a still open PR

BSchilperoort commented on src/ewatercycle/esmvaltool/diagnostic/copier.py:
```diff
@@ -1,13 +1,19 @@
-    get_diagnostic_filename,
-    run_diagnostic,
-)
+try:
```
Moving this to where it's used (i.e. `def *Forcing.generate(...)`) would allow users to mostly use ewatercycle without esmvaltool.

BSchilperoort reviewed (approved):
Tried out this branch. Encountered some issues.

<hr>

1. Tried installing ewatercycle into python (no conda or esmvaltool). Got the following error:

<details><summary>image</summary>

![image](https://github.com/user-attachments/assets/b7bcc6cd-e44d-4943-ae9b-6824892d8492)

</details>

Importlib changed with :snake: 3.10. Changing line 4 in `ewatercycle/shared.py` to:
```py
from importlib.metadata import EntryPoint
```
fixes this issue.

<hr>

2. currently `import ewatercycle.forcing` fails without esmvalcore/tool. This does not have to be this way. Want to fix that in a different PR? Would be nice to open an issue about it.

The import try/excepts can be moved to the methods which rely on esmvaltool, but still allow already generated forcing to be loaded (or e.g. caravanforcing which does not need esmvaltool)

<hr>

Otherwise the conda-lock file seems to work well. It's also a lot faster, so stability + speed are both very nice to have.

sverhoeven committed:
Use importlib from standard library

sverhoeven committed:
On ci use python version that is in lock file

sverhoeven commented:
1. Using importlib from standard library
2. Delayed optonal esmvaltool to #447

sverhoeven committed:
Group importlib with other standard libs

sverhoeven commented:
https://ewatercycle.readthedocs.io/en/stable/system_setup.html#conda-environment) is still using `wget https://raw.githubusercontent.com/eWaterCycle/ewatercycle/main/environment.yml` which will fail. Can be resolved by making a new release.

sverhoeven commented:
Fixes #357
